crest/Assets/Crest/Crest/Shaders/Resources/FFT/FFTSpectrum.compute

```diff
@@ -0,0 +1,217 @@
+// Crest Ocean System
+
+// Copyright 2021 Wave Harmonic Ltd
+
+// Inspired by https://github.com/speps/GX-EncinoWaves
+
+#pragma kernel SpectrumInitalize
+#pragma kernel SpectrumUpdate
+
+#define INV2PI	0.15915494309f
+#define PI4		0.33661977236f
+#define INVPI2	0.63661977236f
+#define HPI		1.57079632679f
+#define PI		3.14159265358f
+#define PI2		6.28318530717f
+#define HSQRT2	0.70710678118f
+
+uint _Size;
+float _WindSpeed;
+float _Turbulence;
+float _Gravity;
+
+uint _RngState;
+
+uint WangHash(uint seed)
+{
+	seed = (seed ^ 61) ^ (seed >> 16);
+	seed *= 9;
+	seed = seed ^ (seed >> 4);
+	seed *= 0x27d4eb2d;
+	seed = seed ^ (seed >> 15);
+	return seed;
+}
+
+uint Rand()
+{
+	_RngState ^= (_RngState << 13);
+	_RngState ^= (_RngState >> 17);
+	_RngState ^= (_RngState << 5);
+	return _RngState;
+}
+
+float RandFloat()
+{
+	return Rand() / 4294967296.0f;
+}
+
+float RandGauss()
+{
+	float u1 = RandFloat();
+	float u2 = RandFloat();
+	if (u1 < 1e-6f)
+		u1 = 1e-6f;
+	return sqrt(-2.0f * log(u1)) * cos(PI2 * u2);
+}
+
+void DeepDispersion(float k, out float w, out float dwdk)
+{
+	w = sqrt(abs(_Gravity * k));
+	dwdk = _Gravity / (2.0f * w);
+}
+
+float AlphaBetaSpectrum(float A, float B, float g, float w, float wm)
+{
+	return
+		(A * g * g / pow(w, 5.0f)) *
+		exp(-B * pow(wm / w, 4.0f));
+}
+
+float PiersonMoskowitzSpectrum(float w)
+{
+	float wm = 0.87f * _Gravity / _WindSpeed;
+	return AlphaBetaSpectrum(8.1e-3f, 1.291f, _Gravity, w, wm);
+}
+
+float PiersonMoskowitzWindTerm( float w )
+{
+	float wm = 0.87f * _Gravity / _WindSpeed;
+	return exp( -1.291 * pow( wm / w, 4.0f ) );
+}
+
+float PosCosSquaredDirectionalSpreading(float w, float theta, float kMag, float dTheta)
+{
+	if (theta > -HPI && theta < HPI)
+	{
+		float ct = cos(theta);
+		return INVPI2 * (ct * ct) * (1 - _Turbulence) + PI4 * _Turbulence;
+	}
+	else
+	{
+		return PI4 * _Turbulence;
+	}
+}
+
+RWTexture2DArray<float4> _ResultInit;
+Texture2D<float> _SpectrumControls;
+SamplerState linear_clamp_sampler;
+
+[numthreads(8,8,1)]
+void SpectrumInitalize(uint3 id : SV_DispatchThreadID)
+{
+	const int2 center = _Size.xx / 2;
+	const int2 coord = id.xy - center;
+
+	uint depth;
+	{
+		uint width, height;
+		_ResultInit.GetDimensions( width, height, depth );
+	}
+
+	if( id.z < (depth - 1) && max( abs( coord.x ), abs( coord.y ) ) < _Size / 4 )
```

Comment:
Metal complains here about mismatching signs as it assumes `depth - 1` is signed.

Reply:
missed that before. i think i've pushed a fix now